tdh/data_source_backup.go

```diff
@@ -0,0 +1,270 @@
+									Computed:    true,
+								},
+								"cluster_size": schema.StringAttribute{
+									Description: "Size of the Instance Type.",
+									Computed:    true,
+								},
+								"backup_location": schema.StringAttribute{
+									Description: "Backup Location",
+									Computed:    true,
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+}
+
+func (d backupDataSource) Read(ctx context.Context, req datasource.ReadRequest, resp *datasource.ReadResponse) {
+	//TODO implement me
+	var state backupResponseModel
+	var backupList []backupResourceModel
+	//tflog.Debug(ctx, "rabbitmq dto", map[string]interface{}{"dto": clusterList})
+	tflog.Debug(ctx, "Service type list :", map[string]interface{}{"dto": state.ServiceType})
+	query := controller.BackupQuery{
+		ServiceType: state.ServiceType.ValueString(),
+	}
+	resp.Diagnostics.Append(req.Config.Get(ctx, &state)...)
+	if err := service_type.ValidateRoleType(state.ServiceType.ValueString()); err != nil {
+		resp.Diagnostics.AddError(
+			"invalid type",
+			err.Error())
+		return
+	}
+
+	backups, err := d.client.Controller.GetBackups(query)
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Unable to Read TDH Service Types",
+			err.Error(),
+		)
+		return
+	}
+
+	if backups.Page.TotalPages > 1 {
+		for i := 1; i <= backups.Page.TotalPages; i++ {
+			query.PageQuery.Index = i - 1
+			totalClusters, err := d.client.Controller.GetBackups(query)
+			if err != nil {
+				resp.Diagnostics.AddError(
+					"Unable to Read TDH Clusters",
+					err.Error(),
+				)
+				return
+			}
+
+			for _, mdsClusterDto := range *totalClusters.Get() {
+				cluster := backupResourceModel{
+					ID:   types.StringValue(mdsClusterDto.Id),
+					Name: types.StringValue(mdsClusterDto.Name),
+
+					ClusterId:         types.StringValue(mdsClusterDto.ClusterId),
+					ClusterName:       types.StringValue(mdsClusterDto.ClusterName),
+					ClusterVersion:    types.StringValue(mdsClusterDto.ClusterVersion),
+					ServiceType:       types.StringValue(mdsClusterDto.ServiceType),
+					Size:              types.StringValue(mdsClusterDto.Size),
+					BackupTriggerType: types.StringValue(mdsClusterDto.BackupTriggerType),
+					Provider:          types.StringValue(mdsClusterDto.Provider),
+					Region:            types.StringValue(mdsClusterDto.Region),
+					DataPlaneId:       types.StringValue(mdsClusterDto.DataPlaneId),
+				}
+
+				metadata := BackupMetadata{
+					ClusterName:    types.StringValue(mdsClusterDto.Metadata.ClusterName),
+					ClusterSize:    types.StringValue(mdsClusterDto.Metadata.ClusterSize),
+					BackupLocation: types.StringValue(mdsClusterDto.Metadata.BackupLocation),
+				}
+				cluster.Metadata = metadata
+				backupList = append(backupList, cluster)
+			}
+		}
+
+		tflog.Debug(ctx, "rabbitmq dto", map[string]interface{}{"dto": backupList})
+		state.BackupList = append(state.BackupList, backupList...)
+	} else {
+		for _, mdsClusterDto := range *backups.Get() {
```

Comment:
here also, keep it just `dto` context free and copy-paste safe, since our code is open sourced, it will reflect bad if not kept consistent.

Reply:
changed